src/ioctls/common/kvm_vec.rs

```diff
@@ -0,0 +1,549 @@
+        let array_size_in_bytes = (mem_allocator_len - 1) * size_of::<T>();
+        array_size_in_bytes / size_of::<T::Entry>()
+    }
+
+    /// Constructs a new KvmVec<T> that contains `num_elements` empty elements
+    /// of type `KvmArray::Entry`
+    ///
+    /// # Arguments
+    ///
+    /// * `num_elements` - The number of empty elements of type `KvmArray::Entry` in the initial `KvmVec`
+    ///
+    pub fn new(num_elements: usize) -> KvmVec<T> {
+        let required_mem_allocator_capacity =
+            KvmVec::<T>::kvm_vec_len_to_mem_allocator_len(num_elements);
+
+        let mut mem_allocator = Vec::with_capacity(required_mem_allocator_capacity);
+        for _ in 0..required_mem_allocator_capacity {
+            mem_allocator.push(T::default())
+        }
+        mem_allocator[0].set_len(num_elements);
+
+        KvmVec {
+            mem_allocator,
+            len: num_elements,
+            capacity: num_elements,
+        }
+    }
+
+    /// Creates a new `KvmVec` structure based on a supplied vector of `KvmArray::Entry`.
+    ///
+    /// # Arguments
+    ///
+    /// * `entries` - The vector of `KvmArray::Entry` entries.
+    ///
+    pub fn from_entries(entries: &[T::Entry]) -> KvmVec<T> {
+        let mut kvm_vec = KvmVec::<T>::new(entries.len());
+
+        {
+            let kvm_vec_entries = kvm_vec.as_mut_kvm_struct().entries_mut();
+            // this is safe because the provided length is correct
+            let kvm_vec_entries_slice = unsafe { kvm_vec_entries.as_mut_slice(entries.len()) };
+            kvm_vec_entries_slice.copy_from_slice(entries);
+        }
+
+        kvm_vec
+    }
+
+    /// Get a reference to the actual KVM structure instance.
+    ///
+    pub fn as_kvm_struct(&self) -> &T {
+        &self.mem_allocator[0]
+    }
+
+    /// Get a mut reference to the actual KVM structure instance.
+    ///
+    pub fn as_mut_kvm_struct(&mut self) -> &mut T {
+        &mut self.mem_allocator[0]
+    }
+
+    /// Get a pointer to the KVM struct so it can be passed to the kernel.
+    ///
+    pub fn as_ptr(&self) -> *const T {
+        self.as_kvm_struct()
+    }
+
+    /// Get a mutable pointer to the KVM struct so it can be passed to the kernel.
+    ///
+    pub fn as_mut_ptr(&mut self) -> *mut T {
+        self.as_mut_kvm_struct()
+    }
+
+    /// Get a mut `Vec<KvmArray::Entry>` that contains all the elements.
+    /// It is important to call `mem::forget` after using this vector.
+    /// Otherwise rust will destroy it.
+    ///
+    fn as_vec(&mut self) -> Vec<T::Entry> {
+        unsafe {
+            let entries_ptr = self.as_mut_kvm_struct().entries_mut().as_mut_ptr();
+            // This is safe since self.len and self.capacity should be correct
+            Vec::from_raw_parts(entries_ptr, self.len, self.capacity as usize)
+        }
+    }
+
+    /// Get the mutable elements slice so they can be modified before passing to the VCPU.
+    ///
+    pub fn as_entries_slice(&self) -> &[T::Entry] {
```

Comment:
I think this can be called `as_slice`, since there no other useful slice-like view into this other than a slice of entries. It's also shorter, and more in line with `Vec` method naming.